tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,88 @@
+#ifndef TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+#define TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+
+#include <stddef.h>
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/constraint_system/variable.h"
+#include "tachyon/zk/air/expressions/air_evaluator.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class ConstraintSystem {
+ public:
+  typedef math::RowMajorMatrix<F> RowMajorMatrix;
+  typedef Eigen::Block<const RowMajorMatrix> ConstMatrixBlock;
+
+  ConstraintSystem() = default;
+  ConstraintSystem(size_t num_public_values, size_t main_width,
+                   size_t preprocessed_width)
+      : num_public_values_(num_public_values),
+        main_width_(main_width),
+        preprocessed_width_(preprocessed_width) {}
+
+  void EnforceFirstRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::FirstRow(std::move(constraint)));
+  }
+
+  void EnforceLastRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::LastRow(std::move(constraint)));
+  }
+
+  void EnforceTransitionConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::Transition(std::move(constraint)));
+  }
+
+  void EnforceConstraint(std::unique_ptr<Expression<F>> constraint) {
+    constraints_.push_back(std::move(constraint));
+  }
+
+  bool IsSatisfied(AirEvaluator<F>& evaluator,
+                   const std::vector<F>& public_values,
+                   const RowMajorMatrix& main,
+                   const RowMajorMatrix* preprocessed = nullptr) const {
+    CHECK_EQ(public_values.size(), num_public_values_);
+    CHECK_EQ(main.cols(), main_width_);
+    if (preprocessed) {
+      CHECK_EQ(preprocessed->cols(), preprocessed_width_);
+      CHECK_EQ(preprocessed->rows(), main.rows());
+    } else {
+      CHECK_EQ(preprocessed_width_, 0);
+    }
+
+    evaluator.set_public_values(public_values);
+    evaluator.set_num_rows(main.rows());
+
+    for (Eigen::Index i = 0; i < main.rows() - 1; ++i) {
+      const ConstMatrixBlock main_window(main, i, 0, 2, main_width_);
+      const ConstMatrixBlock* prep_window = nullptr;
+
+      if (preprocessed) {
+        prep_window =
+            new ConstMatrixBlock(*preprocessed, i, 0, 2, preprocessed_width_);
+      }
+      evaluator.SetCurrentWindowData(i, &main_window, prep_window);
```

Comment:
```suggestion
      if (preprocessed) {
        ConstMatrixBlock prep_window(*preprocessed, i, 0, 2, preprocessed_width_);      
        evaluator.SetCurrentWindowData(i, &main_window, prep_window);
      } else {
        evaluator.SetCurrentWindowData(i, &main_window, nullptr);
      }
```

Reply:
`prep_window` will be destructed outside of the `if` so it will cause UB on the `Evaluate` loop.

Reply:
Oh I see. then could we allocate heap just once and reuse it?